### PR TITLE
Adding TPC-H 0.01 as DuckDB/Arrow correctness test

### PR DESCRIPTION
**Pull request by pdet**
This PR adds:
* TPC-H 0.01 as DuckDB/Arrow correctness test, normal, batched, and parallel (#1720)
* 128-bit Decimal Conversion (#1742)
* Progress Bars and Cardinality Estimation for Arrow Tables
* Refactoring and join-bug fixes.

### CODE REVIEW

File: tools/pythonpkg/tests/arrow/tpch/queries/q01.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
+SELECT
```
Comment (hannesmuehleisen) [inline tools/pythonpkg/tests/arrow/tpch/queries/q01.sql:1]: Same for the queries, could we not export those from the tpch extension to Python, too?

**Review by hannes — CHANGES_REQUESTED** [review]
Thanks, some comments inline

File: src/common/arrow_duckdb.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/assert.hpp"
+#include "duckdb/common/arrow_duckdb.hpp"
+namespace duckdb {
+
+ArrowSchemaDuck::~ArrowSchemaDuck() {
```
Comment (hannesmuehleisen) [inline src/common/arrow_duckdb.cpp:6]: The Duck suffix there seems superfluous

File: src/common/progress_bar.cpp [hunk 3]
```diff
@@ -25,11 +25,16 @@ int ProgressBar::GetCurrentPercentage() {
 
 void ProgressBar::Start() {
 #ifndef DUCKDB_NO_THREADS
```
Comment (hannesmuehleisen) [inline src/common/progress_bar.cpp:27]: Maybe the progress bar changes could be a separate PR?

File: src/include/duckdb/common/arrow_duckdb.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+// duckdb/common/arrow_duckdb.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include "duckdb/common/arrow.hpp"
+//! Here we have the internal duckdb classes that interact with Arrow's Internal Header (i.e., duckdb/commons/arrow.hpp)
+namespace duckdb {
+class ArrowSchemaDuck {
```
Comment (hannesmuehleisen) [inline src/include/duckdb/common/arrow_duckdb.hpp:13]: Same issue with the Duck suffix here

File: test/api/test_arrow.cpp [hunk 5]
```diff
@@ -1,133 +0,0 @@
-#include "catch.hpp"
```
Comment (hannesmuehleisen) [inline test/api/test_arrow.cpp:1]: Why was this deleted without replacement?

File: tools/pythonpkg/tests/arrow/tpch/answers/q01.csv [hunk 6]
```diff
@@ -0,0 +1,5 @@
+l_returnflag|l_linestatus|sum_qty|sum_base_price|sum_disc_price|sum_charge|avg_qty|avg_price|avg_disc|count_order
```
Comment (hannesmuehleisen) [inline tools/pythonpkg/tests/arrow/tpch/answers/q01.csv:1]: Could we not duplicate queries and answers here but export them to Python from the `tpch`  extension?

**Review by hannes — CHANGES_REQUESTED** [review]

File: test/api/test_arrow.cpp (hunk 5 above)
Comment (pdet) [reply test/api/test_arrow.cpp:1]: Yes, these tests were a bit annoying because any changes on the ArrowArrayStream Interface would have to be duplicated there. 
This also just tests the API right, has nothing to do with arrow itself.
so I felt that this is sufficiently covered by the python tests.

File: src/main/query_result.cpp [hunk 7]
```diff
@@ -95,6 +95,8 @@ struct DuckDBArrowSchemaHolder {
 	vector<ArrowSchema> children = {};
 	// unused in children
 	vector<ArrowSchema *> children_ptrs = {};
+	//! This holds strings created to represent decimal types
+	vector<unique_ptr<char[]>> decimal_str;
```
Comment (Mytherin) [inline src/main/query_result.cpp:99]: Perhaps call this something more generic/descriptive, e.g. `owned_type_names`?

File: tools/pythonpkg/src/include/duckdb_python/arrow_array_stream.hpp [hunk 8]
```diff
@@ -9,35 +9,35 @@
 #pragma once
 
 #include <string>
+#include <atomic>
```
Comment (Mytherin) [inline tools/pythonpkg/src/include/duckdb_python/arrow_array_stream.hpp:12]: duckdb/common/atomic.hpp

File: src/function/table/arrow.cpp [hunk 9]
```diff
@@ -240,11 +189,50 @@ void ArrowToDuckDB(ArrowScanState &scan_state, DataChunk &output) {
...
+			break;
+		}
+		case LogicalTypeId::DECIMAL: {
+			//! We have to convert from INT128
+			switch (output.data[col_idx].GetType().InternalType()) {
+			case PhysicalType::INT16: {
+				auto src_ptr = (hugeint_t *)array.buffers[1] + scan_state.chunk_offset + array.offset;
+				auto tgt_ptr = (int16_t *)FlatVector::GetData(output.data[col_idx]);
+				for (idx_t row = 0; row < output.size(); row++) {
+					Hugeint::TryCast(src_ptr[row], tgt_ptr[row]);
```
Comment (Mytherin) [inline src/function/table/arrow.cpp:206]: What happens if ::TryCast returns false? If it somehow can't happen (e.g. this is enforced higher up), then this should at least be asserted.

File: src/function/table/arrow.cpp [hunk 10]
```diff
@@ -334,21 +314,21 @@ bool ArrowScanParallelStateNext(ClientContext &context, const FunctionData *bind
...
 	state.chunk_offset = 0;
-	state.ReleaseArray();
-	auto current_chunk = make_unique<ArrowArray>();
-	if (bind_data.stream->get_next(bind_data.stream, current_chunk.get(), state.chunk_idx)) {
-		throw InvalidInputException("arrow_scan: get_next failed(): %s",
-		                            string(bind_data.stream->get_last_error(bind_data.stream)));
-	}
-	state.chunk = move(current_chunk);
+	state.chunk = bind_data.stream->GetNextChunk();
+	//	auto current_chunk = make_unique<ArrowArray>();
```
Comment (Mytherin) [inline src/function/table/arrow.cpp:318]: Remove the commented out code please

File: src/include/duckdb/function/table/arrow.hpp [hunk 11]
```diff
@@ -9,11 +9,83 @@
 #pragma once
 
 #include "duckdb/function/table_function.hpp"
+#include "duckdb/parallel/parallel_state.hpp"
+#include "duckdb/common/arrow_duckdb.hpp"
+#include <atomic>
```
Comment (Mytherin) [inline src/include/duckdb/function/table/arrow.hpp:14]: duckdb/common/atomic.hpp

File: tools/pythonpkg/src/include/duckdb_python/pyconnection.hpp [hunk 12]
```diff
@@ -24,6 +24,7 @@ struct DuckDBPyConnection {
 	shared_ptr<DuckDB> database;
 	unique_ptr<Connection> connection;
 	unordered_map<string, py::object> registered_dfs;
+	unordered_map<string, py::object> registered_arrow;
```
Comment (Mytherin) [inline tools/pythonpkg/src/include/duckdb_python/pyconnection.hpp:27]: Perhaps we can handle this in a nicer way, e.g. have an abstract class `RegisteredObject` here, that can be overloaded for the DataFrame and the Arrow object separately:

```cpp
class RegisteredObject {
public:
   RegisteredObject(py::obj obj_p) : obj(move(obj_p)) {}
   virtual ~RegisteredObject(){}

   py::obj obj;

public:
   virtual void Unregister() = 0;
};
```

Then we can use a single unordered map here (and potentially unregistering could be done automatically on destruction?).

File: tools/pythonpkg/src/pyconnection.cpp [hunk 13]
```diff
@@ -166,15 +169,18 @@ DuckDBPyConnection *DuckDBPyConnection::RegisterDF(const string &name, py::objec
...
 	}
 	if (table.is_none() || string(py::str(table.get_type().attr("__name__"))) != "Table") {
 		throw std::runtime_error("Only arrow tables supported");
 	}
-	auto stream_factory = make_unique<PythonTableArrowArrayStreamFactory>(table);
-	ArrowArrayStream *(*stream_factory_produce)(uintptr_t factory) = PythonTableArrowArrayStreamFactory::Produce;
+	auto stream_factory = make_unique<PythonTableArrowArrayStreamFactory>(table.ptr());
+	registered_arrow[name] = table;
+
+	unique_ptr<ArrowArrayStreamDuck> (*stream_factory_produce)(uintptr_t factory) =
```
Comment (Mytherin) [inline tools/pythonpkg/src/pyconnection.cpp:182]: auto

File: tools/pythonpkg/tests/arrow/test_progress.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    def test_progress_arrow(self, duckdb_cursor):
+        if not can_run:
+            return
+
+        data = (pyarrow.array(np.arange(10000000), type=pyarrow.int32()))
+        duckdb_conn = duckdb.connect()
+        duckdb_conn.execute("PRAGMA set_progress_bar_time=1")
+        duckdb_conn.execute("PRAGMA disable_print_progress_bar")
+
+        tbl = pyarrow.Table.from_arrays([data],['a'])
```
Comment (Mytherin) [inline tools/pythonpkg/tests/arrow/test_progress.py:23]: Perhaps also test with an empty table?

File: tools/pythonpkg/tests/arrow/test_progress.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        rel = duckdb_conn.from_arrow_table(tbl)
```
Comment (Mytherin) [inline tools/pythonpkg/tests/arrow/test_progress.py:24]: Could you add some tests for the decimals, at least one for each of the different physical types (int16, int32, int64, int128), and also trying to trigger the failure case in Hugeint::TryCast?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Some more feedback:

File: src/common/arrow_duckdb.cpp (hunk 2 above)
Comment (Mytherin) [reply src/common/arrow_duckdb.cpp:6]: Perhaps `Wrapper` is more apt?

File: tools/pythonpkg/tests/arrow/test_progress.py (hunk 15 above)
Comment (pdet) [reply tools/pythonpkg/tests/arrow/test_progress.py:24]: Not related to the test_progress itself right.

Comment (Mytherin) [reply tools/pythonpkg/tests/arrow/test_progress.py:24]: Indeed, just a general request for tests

File: src/common/arrow_duckdb.cpp (hunk 2 above)
Comment (pdet) [reply src/common/arrow_duckdb.cpp:6]: Yeah the problem is that ArrowSchema already exists, I can use wrapper instead of duck, I think makes more sense

File: tools/pythonpkg/tests/arrow/test_progress.py (hunk 15 above)
Comment (pdet) [reply tools/pythonpkg/tests/arrow/test_progress.py:24]: I don't think it is possible to trigger a failure in Hugeint::TryCast (at least shouldn't). I guess you are just suggesting to create decimals on precisions that will be transformed to int16,32 and 64, and be sure that the 128 conversions won't explode there.
Is that it or did I miss something?

File: src/common/progress_bar.cpp (hunk 3 above)
Comment (pdet) [reply src/common/progress_bar.cpp:27]: Yeah, I got carried away, should have done this in a different PR, I think right now is a bit annoying to remove it, but I'll keep in mind to have shorter/smaller PRs in the future. Sorry for that.

File: tools/pythonpkg/tests/arrow/test_progress.py (hunk 15 above)
Comment (Mytherin) [reply tools/pythonpkg/tests/arrow/test_progress.py:24]: Yes exactly. If it should not be possible to trigger such a failure then that's fine as well, as long as there is an assertion there that at least checks the return value of the TryCast.

**Comment by Mytherin** [comment]
Thanks for the changes! Looks good to me now.

**Comment by hannesmuehleisen** [comment]
Thanks!
